FWCore/Concurrency/src/WaitingTaskWithArenaHolder.cc

```diff
@@ -30,7 +30,9 @@ namespace edm {
 
   WaitingTaskWithArenaHolder::WaitingTaskWithArenaHolder(WaitingTaskWithArenaHolder const& iHolder)
       : m_task(iHolder.m_task), m_arena(iHolder.m_arena) {
-    m_task->increment_ref_count();
+    if (m_task) {
+      m_task->increment_ref_count();
+    }
```

Comment:
I cannot remember if this was intentional or not. Probably I just copied this part from WaitingTaskHolder. A similar issue exists with WaitingTaskHolder and also the other constructors. One question would be whether we want to spend the CPU for this extra check in a code path that gets called very frequently. This is in code we are trying to keep lightweight and fast to minimize the overhead of the multithreading infrastructure. And it prevents a failure from a somewhat unusual code path (currently the problem it protects against never happens or else we'd be seeing these seg faults). This is also in an interface that mostly will only be used by the Core group, not by general developers. On the other hand maybe the performance penalty is negligible ...

Reply:
Fair point, thanks.

The `WaitingTaskWithArenaHolder` gets exposed to general developers though via ExternalWork's `acquire()`. And while mostly copying/moving it around and calling `doneWaiting()` should be sufficient, there are cases where one wants to have a chain of CPU tasks and external work. For CUDA (etc) I'm trying to hide the details, but the general issue is there. Although I don't think there is a real need for copying a default-constructed `WaitingTaskWithArenaHolder` (I got hit by the segfault when trying to figure out where my holder got prematurely destructed).

Reply:
Wrap the `if` in `LIKELY`.

Reply:
Thanks @Dr15Jones, done.

What about the `WaitingTaskHolder` that @wddgit pointed out? Would it be worth to add a similar check there?

Additional thought: if we're really concerned about the overhead of added `if` in `WaitingTaskWithArenaHolder`, I think we should consider inlining (at least) all the constructors and assignment operators.

Reply:
@Dr15Jones @makortel do you prefer to integrate this PR as it is (as I would assume given the signature by Chris) and in case address the concerns expressed by @wddgit separately?

Reply:
Please integrate it as it is. I hope at least part of the speed concern is addressed via the use of `LIKELY`.

Reply:
@Dr15Jones ok, I will integrate it in next IB